lang/src/org/partiql/lang/planner/transforms/AstToLogicalVisitorTransform.kt

```diff
@@ -81,6 +86,62 @@ internal class AstToLogicalVisitorTransform(
         return expr
     }
 
+    private fun generateUniqueWindowName(node: PartiqlAst.Expr.CallWindow, index: Int): String {
+        val uniqueId = "\$__partiql_window_function_$index"
+        windowFunctionUniqueIdMap[node] = uniqueId
+        return uniqueId
+    }
+
+    private fun getUniqueWindowName(node: PartiqlAst.Expr.CallWindow): String {
+        return windowFunctionUniqueIdMap[node] ?: error("no such window function registered")
+    }
+
+    /**
+     * If the Select Clause contains window function, create a corresponding Window operator.
+     */
+    private fun extractWindowFunction(node: PartiqlAst.Expr.Select, algebra: PartiqlLogical.Bexpr): PartiqlLogical.Bexpr {
+        // check to see if there is any window call in the select list
+        val windowExpressions = when (val project = node.project) {
+            is PartiqlAst.Projection.ProjectValue -> return algebra
+            is PartiqlAst.Projection.ProjectList -> {
+                project.projectItems.asSequence().filterNot {
+                    it is PartiqlAst.ProjectItem.ProjectAll
+                }.filter {
+                    (it as PartiqlAst.ProjectItem.ProjectExpr).expr is PartiqlAst.Expr.CallWindow
+                }.map {
+                    (((it as PartiqlAst.ProjectItem.ProjectExpr).expr) as PartiqlAst.Expr.CallWindow)
+                }
+            }
+
+            is PartiqlAst.Projection.ProjectPivot -> return algebra
+            is PartiqlAst.Projection.ProjectStar -> return algebra
+        }
+
+        // todo: if multiple window functions are operating on the same window, we can potentially add them in a single window operator to increase performance
+        var modifiedAlgebra = algebra
+        windowExpressions.forEachIndexed { index, callWindow ->
+            modifiedAlgebra = callWindow.let { callWindowNode ->
+                PartiqlLogical.build {
+                    window(
+                        algebra,
```

Comment:
Is this supposed to be `modifiedAlgebra`? It seems like we only really grab the last window expression.